File: apps/web/src/lib/maps/basemap.test.ts
import { describe, expect, it } from "vitest";

import {
  isUnlicensedTileHost,
  NO_BASEMAP_NOTICE,
  readBasemap,
} from "./basemap";

/**
 * #428 — the tile source.
 *
 * The whole module is one rule: WE FAIL TOWARD NO BASEMAP, NEVER TOWARD SOMEBODY
 * ELSE'S GOODWILL. Most of these tests are that rule asked from a different angle,
 * because the failure they prevent is silent — a map that draws is
 * indistinguishable from a map that draws *legally*.
 */

const LICENSED = {
  NEXT_PUBLIC_MAP_TILE_URL: "https://tiles.example.com/{z}/{x}/{y}.png?key=abc",
  NEXT_PUBLIC_MAP_TILE_ATTRIBUTION: "© Example Maps © OpenStreetMap contributors",
};

describe("readBasemap — a configured provider", () => {
  it("returns the URL and attribution together", () => {
    const basemap = readBasemap(LICENSED);
    expect(basemap?.url).toBe(LICENSED.NEXT_PUBLIC_MAP_TILE_URL);
    expect(basemap?.attribution).toContain("Example Maps");
    expect(basemap?.maxZoom).toBe(19);
  });

  it("honours a provider's own max zoom", () => {
    expect(
      readBasemap({ ...LICENSED, NEXT_PUBLIC_MAP_TILE_MAX_ZOOM: "22" })?.maxZoom,
    ).toBe(22);
  });

  it("ignores a nonsense max zoom rather than rendering a broken map", () => {
    for (const value of ["", "abc", "0", "-5"]) {
      expect(
        readBasemap({ ...LICENSED, NEXT_PUBLIC_MAP_TILE_MAX_ZOOM: value })?.maxZoom,
        value,
      ).toBe(19);
    }
  });
});

describe("readBasemap — every unconfigured shape is NO basemap", () => {
  it("returns null when nothing is set", () => {
    expect(readBasemap({})).toBeNull();
  });

  it("returns null for a URL with no attribution", () => {
    // A tile source with no credit is the same licensing problem wearing a
    // different provider's name, so half-configured is treated as unconfigured.
    expect(
      readBasemap({ NEXT_PUBLIC_MAP_TILE_URL: LICENSED.NEXT_PUBLIC_MAP_TILE_URL }),
    ).toBeNull();
  });

  it("returns null for an attribution with no URL", () => {
    expect(
      readBasemap({
        NEXT_PUBLIC_MAP_TILE_ATTRIBUTION: LICENSED.NEXT_PUBLIC_MAP_TILE_ATTRIBUTION,
      }),
    ).toBeNull();
  });

  it("treats whitespace-only values as absent", () => {
    expect(
      readBasemap({
        NEXT_PUBLIC_MAP_TILE_URL: "   ",
        NEXT_PUBLIC_MAP_TILE_ATTRIBUTION: "  ",
      }),
    ).toBeNull();
  });
});

describe("readBasemap — the OSM host cannot be configured back in", () => {
  it("refuses tile.openstreetmap.org even with perfect attribution", () => {
    // This is the point of the module. The violation has to be impossible to
    // reintroduce quietly — by a copied .env, a tutorial, or a well-meant
    // "restore the map" fix — not merely removed once.
    expect(
      readBasemap({
        NEXT_PUBLIC_MAP_TILE_URL: "https://tile.openstreetmap.org/{z}/{x}/{y}.png",
        NEXT_PUBLIC_MAP_TILE_ATTRIBUTION: "© OpenStreetMap contributors",
      }),
    ).toBeNull();
  });

  it("refuses every subdomain form a Leaflet template uses", () => {
    for (const url of [
      "https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png",
      "https://a.tile.openstreetmap.org/{z}/{x}/{y}.png",
      "https://b.tile.openstreetmap.org/1/2/3.png",
      "http://tile.openstreetmap.org/{z}/{x}/{y}.png",
      "https://TILE.OPENSTREETMAP.ORG/{z}/{x}/{y}.png",
    ]) {
      expect(isUnlicensedTileHost(url), url).toBe(true);
    }
  });

  it("does not refuse a lookalike domain that merely contains the string", () => {
    // Over-blocking would be its own bug: a provider whose hostname happens to
    // contain the phrase is licensed and must work.
    for (const url of [
      "https://tiles.example.com/tile.openstreetmap.org.png",
      "https://tile.openstreetmap.org.evil.example.com/{z}/{x}/{y}.png",
      "https://my-tile-openstreetmap-org.example.com/{z}/{x}/{y}.png",
    ]) {
      expect(isUnlicensedTileHost(url), url).toBe(false);
    }
  });

  it("still permits a provider that SERVES OSM data under its own terms", () => {
    // OSM *data* is fine and widely resold; it is OSM's donated *tile servers*
    // that are not ours to use. Blocking the data would be the wrong lesson.
    expect(
      isUnlicensedTileHost("https://api.maptiler.com/maps/streets/{z}/{x}/{y}.png?key=k"),
    ).toBe(false);
  });
});

describe("the no-basemap notice", () => {
  it("says the pins are still right, and who can fix the background", () => {
    // The state is a configuration an owner completes, not a fault to apologise
    // for — and the crew needs to know the pin positions are unaffected.
    expect(NO_BASEMAP_NOTICE).toMatch(/pins are exact/i);
    expect(NO_BASEMAP_NOTICE).toMatch(/owner/i);
    expect(NO_BASEMAP_NOTICE).not.toMatch(/error|sorry|failed|broken/i);
  });
});

File: apps/android/app/src/test/kotlin/com/loonext/android/features/tasks/BasemapTest.kt
package com.loonext.android.features.tasks

import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * #428 — the tile source.
 *
 * The regex is a HAND PORT of the web twin's `isUnlicensedTileHost`, and the two
 * failure directions are not symmetric:
 *
 *   MISSING a match reintroduces the licensing violation silently, because a map
 *   that draws looks identical to a map that draws legally.
 *   OVER-matching blocks a provider we ARE licensed to use, which at least
 *   presents as a visibly missing basemap.
 *
 * Both are tested. Kotlin's Regex also treats escapes differently from JS in ways
 * that have bitten this codebase before, so the port is asserted rather than
 * assumed.
 */
class BasemapTest {
    @Test
    fun `refuses every subdomain form a tile template uses`() {
        for (url in listOf(
            "https://tile.openstreetmap.org/{z}/{x}/{y}.png",
            "https://a.tile.openstreetmap.org/{z}/{x}/{y}.png",
            "https://b.tile.openstreetmap.org/1/2/3.png",
            "http://tile.openstreetmap.org/{z}/{x}/{y}.png",
            "https://TILE.OPENSTREETMAP.ORG/{z}/{x}/{y}.png",
        )) {
            assertTrue(url, Basemap.isUnlicensedTileHost(url))
        }
    }

    @Test
    fun `does not refuse a lookalike domain that merely contains the string`() {
        // Over-blocking is its own bug: a licensed provider whose hostname happens
        // to contain the phrase must still work.
        for (url in listOf(
            "https://tiles.example.com/tile.openstreetmap.org.png",
            "https://tile.openstreetmap.org.evil.example.com/{z}/{x}/{y}.png",
            "https://my-tile-openstreetmap-org.example.com/{z}/{x}/{y}.png",
        )) {
            assertFalse(url, Basemap.isUnlicensedTileHost(url))
        }
    }

    @Test
    fun `still permits a provider that serves OSM DATA under its own terms`() {
        // OSM data is fine and widely resold; it is OSM's donated tile SERVERS that
        // are not ours to use. Blocking the data would be the wrong lesson.
        assertFalse(
            Basemap.isUnlicensedTileHost(
                "https://api.maptiler.com/maps/streets/{z}/{x}/{y}.png?key=k",
            ),
        )
    }

    @Test
    fun `ships unconfigured, which is the compliant state`() {
        // The build defaults are empty on purpose. If this ever fails because
        // somebody filled them in, that is fine — but it must be a deliberate act
        // with a provider whose terms permit it, not a default nobody chose.
        assertFalse(Basemap.isConfigured())
    }

    @Test
    fun `the notice matches the web twin verbatim`() {
        // Same sentence on both clients: a crew member who sees it on the phone and
        // then on the laptop should not wonder whether they are two problems.
        assertTrue(Basemap.NO_BASEMAP_NOTICE.contains("Job pins are exact"))
        assertTrue(Basemap.NO_BASEMAP_NOTICE.contains("owner can do in one setting"))
        assertFalse(Basemap.NO_BASEMAP_NOTICE.contains("error"))
    }
}
